javaslang/src/main/java/javaslang/collection/Vector.java

```diff
@@ -1387,23 +1414,376 @@ public String toString() {
+        }
+
+        static <T> T[] getLeaf(Object[] root, int depth, int index) {
+            assert (depth >= 2) && (index >= 0);
+
+            int shift = shiftValue(depth);
+            Object[] array = (Object[]) root[index >> shift]; // doesn't need masking
+
+            while (shift > BRANCHING_BASE) {
+                shift -= BRANCHING_BASE;
+                array = (Object[]) array[(index >> shift) & branchingMask()]; // TODO mask and shift should be hidden in a proper method, e.g. nth
+            }
+
+            assert array != null;
+            return (T[]) array;
+        }
+
+        static int shiftValue(int depth)   { return (depth - 1) * BRANCHING_BASE; }
+        static int fullTreeSize(int depth) { return branchingFactor() * fullBranchSize(depth); }
+
+        /* branchingFactor()^(depth-1) == 2^(BRANCHING_BASE*depth) */
+        static int fullBranchSize(int depth) {
+            final int result = 1 << (BRANCHING_BASE * (depth - 1));
+            assert result == (int) Math.pow(branchingFactor(), (depth - 1));
+            return result;
+        }
     }
-}
 
-interface VectorModule {
+    static final class Leaf<T> implements Serializable { // TODO separate start/end?
+        private static final long serialVersionUID = 1L;
+        private static final Leaf<?> EMPTY = new Leaf<>(emptyArray(), (byte) 0);
 
-    final class Combinations {
+        private final T[] array;
+        private final byte length;
 
-        static <T> Vector<Vector<T>> apply(Vector<T> elements, int k) {
-            if (k == 0) {
-                return Vector.of(Vector.empty());
+        static <T> Leaf<T> empty()                         { return (Leaf<T>) EMPTY; }
+        public static <T> Leaf<T> of(T[] leaf)             { return of(leaf, leaf.length); }
+        public static <T> Leaf<T> of(T[] leaf, int length) { return new Leaf<T>(leaf, (byte) length); }
+
+        private Leaf(T[] array, byte length) {
+            assert array.length <= branchingFactor();
+
+            this.array = array;
+            this.length = (byte) Math.min(length, array.length);
+        }
+
+        T[] start() { return collapse(0, length()); }
+        T[] end()   { return collapse(array.length - length(), array.length); }
+        private T[] collapse(int from, int to) {
+            return ((to - from) == array.length)
+                   ? array
+                   : copyOfRange(array, from, to);
+        }
+
+        boolean isEmpty()        { return length() == 0; }
+
+        Leaf<T> drop(int delta)  { return of(array, length() - delta); }
+        Leaf<T> take(int length) { return of(array, length); }
+
+        int offset()             { return array.length - length(); }
+        byte length()            { return length; }
+
+        T start(int index) {
+            final int offset = offset();
+            assert (offset >= 0) && (offset < array.length);
+            return end(offset + index);
+        }
+        T end(int index) {
```

Comment:
> Also, when does it make sense to use start() instead of end()? When otherwise? 

I figured that one out myself: `start` and `start(int)` are only used on `Vector.firstLeaf`, and `end` and `end(int)` are only used on `Vector.lastLeaf`. It seems to me that the two Vector fields (`firstLeaf` and `lastLeaf`) should have different types, something that only exposes the right element collection (`start` and `end`) and indexing (`start(int)`, `end(int)`) methods.